libcoap/src/context.rs

```diff
@@ -401,6 +425,151 @@ impl CoapContext<'_> {
+        if let Some(initial_recipient) = oscore_conf.initial_recipient.clone() {
+            let initial_recipient = OscoreRecipient::new(initial_recipient.as_str());
+            inner_ref.recipients.push(initial_recipient);
+        }
+
+        // Mark context as valid oscore server.
+        inner_ref.provided_oscore_information = true;
+        Ok(())
+    }
+
+    /// Adds a recipient_id to the OscoreContext.
+    ///
+    /// # Errors
+    /// Will return a [OscoreRecipientError] if adding the recipient failed.
+    /// You might want to check the error to determine if adding the recipient failed
+    /// due to the recipient_id beeing already added to the context.
+    /// Trying to call this on a CoapContext without appropriate oscore information will
+    /// also result in an error.
+    #[cfg(feature = "oscore")]
+    pub fn new_oscore_recipient(&mut self, recipient_id: &str) -> Result<(), OscoreRecipientError> {
+        let mut inner_ref = self.inner.borrow_mut();
+
+        // Return if context has no appropriate oscore information.
+        if !inner_ref.provided_oscore_information {
+            #[cfg(debug_assertions)]
+            eprintln!("tried adding recipient to context with no appropriate oscore information");
+            return Err(OscoreRecipientError::NoOscoreContext);
+        }
+
+        // Return if the recipient_id is already added as the coap_new_oscore_recipient
+        // would free the raw struct for duplicate recipient_id's.
+        // As we can't check why adding the recipient_id failed we have to filter
+        // this case to prevent a double free.
+        if inner_ref
+            .recipients
+            .iter()
+            .any(|elem| elem.get_recipient_id() == recipient_id)
+        {
+            return Err(OscoreRecipientError::DuplicateId);
+        }
+
+        let recipient = OscoreRecipient::new(recipient_id);
+        let result: i32;
+
+        // SAFETY: If adding the recipient to the context fails we drop its underlying raw
+        // struct manually as it's not handled by the raw_context. There is one case where
+        // libcoap would already free the raw struct, which is filtered out above as we
+        // prevent adding a duplicate recipient_id to the context and return.
+        unsafe {
+            result = coap_new_oscore_recipient(inner_ref.raw_context, recipient.get_c_struct());
+        };
+
+        // Drop the raw struct if adding it failed (except for duplicate id)...
+        // SAFETY: This should be safe to use here as 'coap_new_oscore_recipient()' would only
+        // free() the raw struct on failure due to a duplicate recipient_id, which is filtered
+        // out above.
+        if result == 0 {
+            recipient.drop();
+            return Err(OscoreRecipientError::Unknown);
+        }
+
+        // ...or else save the recipient to keep alive the pointer to its raw struct.
+        inner_ref.recipients.push(recipient);
```

Comment:
Raw pointers are not freed if the struct containing them is dropped, so you might not even have to keep the `OscoreRecipient` stored at all.

In fact, you might also get away with not having an `OscoreRecipient` struct at all and just taking a `recipient_id: &str` as arguments for this method and `delete_oscore_recipient()` instead.